### PR TITLE
Improve memoization to deal with diamonds better

### PR DESCRIPTION
**Pull request by johnynek**
In large summingbird graphs, merges can cause exponential runtime in planning because we repeatedly call on the state function on both sides of the merge.

We had a partial solution for this, which worked to make sure the data was not incorrect, but there is a second outer layer: computing what the acceptable time range is. This was the layer that was not memoized, and it was making it take forever to plan for us.

### CODE REVIEW

**Comment by johnynek** [comment]
@ttim can you take a look?

**Comment by ttim** [comment]
@johnynek I'll take a look later today!

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala [hunk 1]
```diff
@@ -262,10 +280,20 @@ object Scalding {
    *  If we memoize with this function, it guarantees that the PipeFactory
    *  is idempotent.
    */
-  def memoize[T](pf: PipeFactory[T]): PipeFactory[T] = {
-    val memo = new Memo[T]
-    pf.map { rdr =>
-      Reader({ i => memo.getOrElseUpdate(i, rdr) })
+  def memoize[T](pf: PlannerOutput[FlowProducer[T]]): PlannerOutput[FlowProducer[T]] = {
```
Comment (ttim) [inline summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala:287]: Can we put `PipeFactor[T]` back here?

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala [hunk 2]
```diff
@@ -252,6 +252,24 @@ object Scalding {
   def also[L, R](ensure: FlowToPipe[L], result: FlowToPipe[R]): FlowToPipe[R] =
     joinFP(ensure, result).map { case (_, r) => r }
 
+  def memoizePlan[A](pf: PlannerOutput[A]): PlannerOutput[A] = {
```
Comment (ttim) [inline summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: I think generic `memoize(StateWithError)` is simpler to understand.

Comment (ttim) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: This function doesn't memoize failures?

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala [hunk 3]
```diff
@@ -262,10 +280,20 @@ object Scalding {
    *  If we memoize with this function, it guarantees that the PipeFactory
    *  is idempotent.
    */
-  def memoize[T](pf: PipeFactory[T]): PipeFactory[T] = {
-    val memo = new Memo[T]
-    pf.map { rdr =>
-      Reader({ i => memo.getOrElseUpdate(i, rdr) })
+  def memoize[T](pf: PlannerOutput[FlowProducer[T]]): PlannerOutput[FlowProducer[T]] = {
+    val memo = new Memo[FlowInput, T]
+    memoizePlan(pf).map { rdr =>
```
Comment (ttim) [inline summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: Is `memoizePlan(pf).map { ... as before ... }` has different semantic with this code?

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala [hunk 4]
```diff
@@ -649,11 +677,22 @@ object Scalding {
...
 
-// Jank to get around serialization issues
-class Memo[T] extends java.io.Serializable {
-  @transient private val mmap = scala.collection.mutable.Map[(FlowDef, Mode), TimedPipe[T]]()
-  def getOrElseUpdate(in: (FlowDef, Mode), rdr: Reader[(FlowDef, Mode), TimedPipe[T]]): TimedPipe[T] =
-    mmap.getOrElseUpdate(in, rdr(in))
+class Memo[A, B] extends java.io.Serializable {
+  // Jank to get around serialization issues
+  @transient private[this] var mmap = Map.empty[A, B]
+  def getOrElseUpdate(a: A, b: => B): B =
```
Comment (ttim) [inline summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: Can be removed

**Review by ttim — CHANGES_REQUESTED** [review]
Unrelated to your change I'm not quite understand comment before `memoize` function. It says you need to do this memoization for correctness but I can't see how. Do we build `PipeFactory` more than once?

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala (hunk 1 above)
Comment (johnynek) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala:287]: It was needlessly constraining the type. This works for any T. `PipeFactory[T] == PlannerOutput[FlowProducer[TimedPipe[T]]]` so I didn't want to limit it.

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala (hunk 2 above)
Comment (johnynek) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: it does not memoize failures.

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala (hunk 3 above)
Comment (johnynek) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: Previously, we didn't memoize the state computation, just the reader inside. The semantics are the same, but this removes an exponential blowup on graphs that merge back into themselves repeatedly (similar to the problem with a naive implementation of fibonacci).

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala (hunk 1 above)
Comment (ttim) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala:287]: Makes sense, looks better this way.

File: summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala (hunk 3 above)
Comment (ttim) [reply summingbird-scalding/src/main/scala/com/twitter/summingbird/scalding/ScaldingPlatform.scala]: Oh, I see. Basically you do `memoizeReader(Reader[I, O]) -> Reader[I, O]`. Then I prefer to have it extracted and then what you want is `memoizeState(pf).map(memoizeReader)`.

**Comment by johnynek** [comment]
@ttim I made the changes you requested. Can you take another look?

**Review by ttim — APPROVED** [review]
I think with `memoizeReader` this would be better but I'm good either way.

**Comment by johnynek** [comment]
Thanks for the review @ttim

**Pull request merged**